core/threaded/scheduler_QS.c

```diff
@@ -0,0 +1,326 @@
+/* Quasi-static scheduler for the threaded runtime of the C target of Lingua
+Franca. */
+
+/*************
+Copyright (c) 2022, The University of Texas at Dallas. Copyright (c) 2022, The
+University of California at Berkeley.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/**
+ * Quasi-static scheduler for the threaded runtime of the C target of Lingua
+ * Franca.
+ *
+ * @author{Shaokai Lin <shaokai@berkeley.edu>}
+ */
+
+#ifndef NUMBER_OF_WORKERS
+#define NUMBER_OF_WORKERS 1
+#endif  // NUMBER_OF_WORKERS
+
+#include <assert.h>
+
+#include "../platform.h"
+#include "../utils/semaphore.h"
+#include "scheduler.h"
+#include "scheduler_instance.h"
+#include "scheduler_sync_tag_advance.c"
+
+#include "../../schedule.h" // Generated
+
+/////////////////// External Variables /////////////////////////
+extern lf_mutex_t mutex;
+extern const int reaction_count;
+extern const int num_semaphores;
+extern const inst_t** static_schedules[];
+extern const uint32_t* schedule_lengths[];
+
+/////////////////// Scheduler Variables and Structs /////////////////////////
+_lf_sched_instance_t* _lf_sched_instance;
+
+/////////////////// Scheduler Private API /////////////////////////
+/**
+ * @brief If there is work to be done, notify workers individually.
+ *
+ * This assumes that the caller is not holding any thread mutexes.
+ */
+void _lf_sched_notify_workers() {    
+    // Calculate the number of workers that we need to wake up, which is the
+    // Note: All threads are idle. Therefore, there is no need to lock the mutex
+    // while accessing the index for the current level.
```

Comment:
It looks like the mutex is held when this is called.

Reply:
I will see if the mutex can be made unlocked when this is called.